protocols/gossipsub/src/behaviour.rs

```diff
@@ -0,0 +1,1165 @@
+    pub fn new(local_peer_id: PeerId, gs_config: GossipsubConfig) -> Self {
+        Gossipsub {
+            config: gs_config.clone(),
+            events: VecDeque::new(),
+            control_pool: HashMap::new(),
+            local_peer_id,
+            topic_peers: HashMap::new(),
+            peer_topics: HashMap::new(),
+            mesh: HashMap::new(),
+            fanout: HashMap::new(),
+            fanout_last_pub: HashMap::new(),
+            mcache: MessageCache::new(gs_config.history_gossip, gs_config.history_length),
+            received: LruCache::new(256), // keep track of the last 256 messages
+            heartbeat: Interval::new(
+                Instant::now() + gs_config.heartbeat_initial_delay,
+                gs_config.heartbeat_interval,
+            ),
+            marker: PhantomData,
+        }
+    }
+
+    /// Subscribes to a topic.
+    ///
+    /// Returns true if the subscription worked. Returns false if we were already subscribed.
+    pub fn subscribe(&mut self, topic: Topic) -> bool {
+        debug!("Subscribing to topic: {}", topic);
+        let topic_hash = self.topic_hash(topic.clone());
+        if self.mesh.get(&topic_hash).is_some() {
+            debug!("Topic: {} is already in the mesh.", topic);
+            return false;
+        }
+
+        // send subscription request to all peers in the topic
+        if let Some(peer_list) = self.topic_peers.get(&topic_hash) {
+            let event = Arc::new(GossipsubRpc {
+                messages: Vec::new(),
+                subscriptions: vec![GossipsubSubscription {
+                    topic_hash: topic_hash.clone(),
+                    action: GossipsubSubscriptionAction::Subscribe,
+                }],
+                control_msgs: Vec::new(),
+            });
+
+            for peer in peer_list {
+                debug!("Sending SUBSCRIBE to peer: {:?}", peer);
+                self.events.push_back(NetworkBehaviourAction::SendEvent {
+                    peer_id: peer.clone(),
+                    event: event.clone(),
+                });
+            }
+        }
+
+        // call JOIN(topic)
+        // this will add new peers to the mesh for the topic
+        self.join(&topic_hash);
+        info!("Subscribed to topic: {}", topic);
+        true
+    }
+
+    /// Unsubscribes from a topic.
+    ///
+    /// Returns true if we were subscribed to this topic.
+    pub fn unsubscribe(&mut self, topic: Topic) -> bool {
+        debug!("Unsubscribing from topic: {}", topic);
+        let topic_hash = &self.topic_hash(topic);
+
+        if self.mesh.get(topic_hash).is_none() {
+            debug!("Already unsubscribed to topic: {:?}", topic_hash);
+            // we are not subscribed
+            return false;
+        }
+
+        // announce to all peers in the topic
+        if let Some(peer_list) = self.topic_peers.get(topic_hash) {
+            let event = Arc::new(GossipsubRpc {
```

Comment:
If I recall correctly, the original idea behind this was to prevent duplicating the memory for each event that have the same message. There is no con-current access, so an `Rc` is better than an `Arc`. 

Originally, the `event` queue just contained the object. I've swapped the `Arc`'s to `Rc` but am open to suggestions as to whether keep this at all, or go back to using the object itself.

Reply:
After talking to @twittner I think we should keep the `Arc`s. In case they ever show up as a performance bottleneck we can still optimize later on. Sorry for the noise.

Reply:
All good. Will keep